### PR TITLE
Capture the `ctrl` transform into jaxpr

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

We have an ongoing experimental project to enable capture of quantum functions into a jaxpr representation.

Following on from #5966 , this PR adds the ability to capture the `ctrl` transform into jaxpr.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-68090]

### CODE REVIEW

File: pennylane/ops/op_math/controlled.py [hunk 1]
```diff
@@ -204,6 +224,62 @@ def wrapper(*args, **kwargs):
...
+        control_wires = args[-n_control:]
+        args = args[:-n_control]
+
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(jaxpr, consts, *args)
+        ops, _ = qml.queuing.process_queue(q)
+
+        for op in ops:
+            ctrl(op, control_wires, control_values, work_wires)
+        return []
```
Comment (dime10) [inline pennylane/ops/op_math/controlled.py:249]: I think here we agreed on `None` right?
```suggestion
```

File: pennylane/ops/op_math/controlled.py [hunk 2]
```diff
@@ -204,6 +224,62 @@ def wrapper(*args, **kwargs):
...
+
+    return ctrl_prim
+
+
+def _capture_ctrl_transform(qfunc: Callable, control, control_values, work_wires) -> Callable:
+    """Capture compatible way of performing an ctrl transform."""
+    # note that this logic is tested in `tests/capture/test_nested_plxpr.py`
+    import jax  # pylint: disable=import-outside-toplevel
+
+    qnode_prim = _get_ctrl_qfunc_prim()
```
Comment (dime10) [inline pennylane/ops/op_math/controlled.py:263]:
```suggestion
    ctrl_prim = _get_ctrl_qfunc_prim()
```

File: pennylane/ops/op_math/controlled.py [hunk 3]
```diff
@@ -204,6 +224,62 @@ def wrapper(*args, **kwargs):
...
+
+    @wraps(qfunc)
+    def new_qfunc(*args, **kwargs):
+        jaxpr = jax.make_jaxpr(functools.partial(qfunc, **kwargs))(*args)
+        control_wires = qml.wires.Wires(control)  # make sure is iterable
+        return qnode_prim.bind(
```
Comment (dime10) [inline pennylane/ops/op_math/controlled.py:269]:
```suggestion
        return ctrl_prim.bind(
```

**Review by dime10 — APPROVED** [review]
Nice :)

File: pennylane/ops/op_math/controlled.py [hunk 4]
```diff
@@ -204,6 +224,62 @@ def wrapper(*args, **kwargs):
...
+    # note that this logic is tested in `tests/capture/test_nested_plxpr.py`
+    import jax  # pylint: disable=import-outside-toplevel
+
+    qnode_prim = _get_ctrl_qfunc_prim()
+
+    @wraps(qfunc)
+    def new_qfunc(*args, **kwargs):
+        jaxpr = jax.make_jaxpr(functools.partial(qfunc, **kwargs))(*args)
+        control_wires = qml.wires.Wires(control)  # make sure is iterable
+        qnode_prim.bind(
```
Comment (dwierichs) [inline pennylane/ops/op_math/controlled.py:269]:
```suggestion
        ctrl_prim.bind(
```

File: pennylane/ops/op_math/controlled.py [hunk 5]
```diff
@@ -204,6 +224,62 @@ def wrapper(*args, **kwargs):
...
+
+    ctrl_prim = jax.core.Primitive("ctrl_transform")
+    ctrl_prim.multiple_results = True
+
+    @ctrl_prim.def_impl
+    def _(*args, n_control, jaxpr, control_values, work_wires, n_consts):
+        consts = args[:n_consts]
+        args = args[n_consts:]
+        control_wires = args[-n_control:]
+        args = args[:-n_control]
```
Comment (dwierichs) [inline pennylane/ops/op_math/controlled.py:241]: This is a very subjective suggestion, but in my head this reads easier:
```suggestion
        consts = args[:n_consts]
        control_wires = args[-n_control:]
        args = args[n_consts:-n_control]
```

File: tests/capture/test_nested_plxpr.py [hunk 6]
```diff
@@ -159,3 +161,111 @@ def qfunc(wire):  # x is closure variable and a tracer
...
+        def f(x, w):
+            return qml.ctrl(qml.RX, 1)(x, w)
+
+        plxpr = jax.make_jaxpr(f)(0.5, 0)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            out = jax.core.eval_jaxpr(plxpr.jaxpr, plxpr.consts, 1.2, 2)
+
+        assert f(0.5, 0) is None
+        assert out == []
```
Comment (dwierichs) [inline tests/capture/test_nested_plxpr.py:181]: Following @dime10's suggestion.
```suggestion
        assert out is None
```

File: tests/capture/test_nested_plxpr.py [hunk 7]
```diff
@@ -159,3 +161,111 @@ def qfunc(wire):  # x is closure variable and a tracer
...
+        expected = qml.ctrl(qml.RX(1.2, 2), 1)
+        qml.assert_equal(q.queue[0], expected)
+
+        assert plxpr.eqns[0].primitive == ctrl_prim
+        assert plxpr.eqns[0].params["control_values"] == [True]
+        assert plxpr.eqns[0].params["n_control"] == 1
+        assert plxpr.eqns[0].params["work_wires"] is None
```
Comment (dwierichs) [inline tests/capture/test_nested_plxpr.py:188]:
```suggestion
        assert plxpr.eqns[0].params["work_wires"] is None
        assert plxpr.eqns[0].params["n_consts"] == 0
```

File: tests/capture/test_nested_plxpr.py [hunk 8]
```diff
@@ -159,3 +161,111 @@ def qfunc(wire):  # x is closure variable and a tracer
...
+
+        def f(w1, w2, w3):
+            return qml.ctrl(qml.X, (w2, w3))(w1)
+
+        plxpr = jax.make_jaxpr(f)(4, 5, 6)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            out = jax.core.eval_jaxpr(plxpr.jaxpr, plxpr.consts, 1, 2, 3)
+
+        assert out == []
```
Comment (dwierichs) [inline tests/capture/test_nested_plxpr.py:201]:
```suggestion
        assert out is None
```

File: tests/capture/test_nested_plxpr.py [hunk 9]
```diff
@@ -159,3 +161,111 @@ def qfunc(wire):  # x is closure variable and a tracer
...
+
+        def f(w):
+            return qml.ctrl(qml.S, (1, 2), work_wires="aux")(w)
+
+        plxpr = jax.make_jaxpr(f)(6)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            out = jax.core.eval_jaxpr(plxpr.jaxpr, plxpr.consts, 5)
+
+        assert out == []
```
Comment (dwierichs) [inline tests/capture/test_nested_plxpr.py:222]:
```suggestion
        assert out is None
```

File: tests/capture/test_nested_plxpr.py [hunk 10]
```diff
@@ -159,3 +161,111 @@ def qfunc(wire):  # x is closure variable and a tracer
...
+
+        def f(z):
+            return qml.ctrl(qml.RZ, (3, 4), [False, True])(z, 0)
+
+        plxpr = jax.make_jaxpr(f)(0.5)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            out = jax.core.eval_jaxpr(plxpr.jaxpr, plxpr.consts, 5.4)
+
+        assert out == []
```
Comment (dwierichs) [inline tests/capture/test_nested_plxpr.py:240]:
```suggestion
        assert out is None
```

File: tests/capture/test_nested_plxpr.py [hunk 11]
```diff
@@ -159,3 +161,111 @@ def qfunc(wire):  # x is closure variable and a tracer
...
+    def test_nested_control(self):
+        """Test that control can be nested."""
+
+        def f(x, w1, w2):
+            f1 = qml.ctrl(qml.Rot, w1)
+            return qml.ctrl(f1, w2)(x, 0.5, 2 * x, 0)
+
+        plxpr = jax.make_jaxpr(f)(-0.5, 1, 2)
+
+        print(plxpr)
```
Comment (dwierichs) [inline tests/capture/test_nested_plxpr.py:257]:
```suggestion
```

File: tests/capture/test_nested_plxpr.py [hunk 12]
```diff
@@ -159,3 +161,111 @@ def qfunc(wire):  # x is closure variable and a tracer
...
+
+        assert out == []
+        expected = qml.ctrl(qml.RZ(5.4, 0), (3, 4), [False, True])
+        qml.assert_equal(q.queue[0], expected)
+        assert len(q) == 1
+
+        assert plxpr.eqns[0].params["control_values"] == [False, True]
+        assert plxpr.eqns[0].params["n_control"] == 2
+
+    def test_nested_control(self):
```
Comment (dwierichs) [inline tests/capture/test_nested_plxpr.py:248]: Do we also need a test case where we pass a callable to `ctrl` that produces multiple operations?

File: tests/capture/test_nested_plxpr.py [hunk 13]
```diff
@@ -159,3 +161,111 @@ def qfunc(wire):  # x is closure variable and a tracer
...
+        """Test that control can be nested."""
+
+        def f(x, w1, w2):
+            f1 = qml.ctrl(qml.Rot, w1)
+            return qml.ctrl(f1, w2)(x, 0.5, 2 * x, 0)
+
+        plxpr = jax.make_jaxpr(f)(-0.5, 1, 2)
+
+        print(plxpr)
+        assert plxpr.eqns[1].params["n_consts"] == 1
```
Comment (dwierichs) [inline tests/capture/test_nested_plxpr.py:258]:
```suggestion
        # First equation of plxpr is the multiplication of x by 2
        assert plxpr.eqns[1].params["n_consts"] == 1 # w1 is a const for the outer `ctrl`
```

File: pennylane/ops/op_math/controlled.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/ops/op_math/controlled.py:249]: The primitives themselves can't return None, as there's no "abstract none".  But the user-facing `qml.ctrl` will return `None` in this case.

File: tests/capture/test_nested_plxpr.py (hunk 6 above)
Comment (albi3ro) [reply tests/capture/test_nested_plxpr.py:181]: `jax.core.eval_jaxpr` always returns a list of tensorlike's.  But we do have `f(0.5, 0) is None` tested on the line above.

File: tests/capture/test_nested_plxpr.py (hunk 8 above)
Comment (albi3ro) [reply tests/capture/test_nested_plxpr.py:201]: Again, can't override the behavior of `jax.core.eval_jaxpr`.

File: tests/capture/test_nested_plxpr.py (hunk 12 above)
Comment (albi3ro) [reply tests/capture/test_nested_plxpr.py:249]: adding.

**Review by dwierichs — APPROVED** [review]
Very nice, thanks @albi3ro!
I got confused with the return statement of the `def_impl` but I think I'm on board now with `[]` :D

**Pull request merged**